### ISSUE TITLE
Proposal: composable seccomp profiles

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#22109]
Docker now has the ability to set seccomp profiles on a container (https://github.com/docker/docker/pull/17989), 
however, writing a custom profile can be tedious, and leads to a lot 
of repetition if a profile only differs slightly from the default profile.

To reduce this repetition, I suggest we allow users to write "partial" profiles,
and combine them to make up for a full profile (by providing multiple
 `--security-opt seccomp=..` flags).

In addition, we can provide some easy to use partials for people to pick from,
and include those in the default install.

Basically;

docker-disable-chown-chmod.json:

```JSON
{
    "defaultAction": "SCMP_ACT_ALLOW",
    "architectures": [
        "SCMP_ARCH_X86_64",
        "SCMP_ARCH_X86",
        "SCMP_ARCH_X32"
    ],
    "syscalls": [
        {
            "name": "chmod",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        },
        {
            "name": "chown",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        },
        {
            "name": "chown32",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        }
    ]
}
```

Run an image, using the "docker-default" profile, and apply a second (partial)
profile that disables "chmod", "chown";

    docker run -d \
      --security-opt seccomp=docker-default.json \
      --security-opt seccomp=disable-chown-chmod.json \
      myimage

### Questions

- Should the default profile always be applied, unless `--security-opt seccomp=unconfined` is used?
- How to handle the `"defaultAction"`? i.e., if `"defaultAction"` is specified in a partial profile, does it change the "overall" default, or is it only used as a default for seccomp-calls inside that file (if no action is specified)?
- Should we simply apply/merge profiles in the order they were specified, or should they be applied as "`deny` trumps `allow`", thus once denied, a profile cannot "allow"?
- Should "partial" templates have a slightly different format, to indicate it's a "partial"?

### Implementation

This depends a bit on the questions above; for a "clean" API, I suggest to perform the merging *client side*, so the client
combines/merges the profiles and sends them as a full profile to the daemon.

*However*, this may be difficult if the default profile should always be applied, because the default profile
is not available to the client. So, either the user should install the `docker-default` profile on the client,
or merging the default profile should be done on the daemon (which will be backwards-incompatible with the current behavior).

**Comment by thaJeztah** [comment]
ping @jfrazelle @justincormack - just something I was thinking about; I think we discussed something like this, but wasn't sure if it was written down somewhere

**Comment by runcom** [comment]
ping @mheon, I believe he's looking into this also

**Comment by justincormack** [comment]
We definitely want more flexibility in composing profiles, but you can't do it quite like this because our default policy is `DENY` not `ALLOW` (correctly, pretty muc any policy starting `ALLOW` can be bypassed using another system call that can do the same thing, eg in your example you miss `chownat` which can do the same as `chown`) and `DENY` policies do not compose as nicely, you cant apply one after the other you need to combine them. Also policies are kind of nasty to write, as they are so long.

Currently I am thinking about extending the default policy to vary by which capabilities you are using, so that eg adding `CAP_SYS_ADMIN` allows `mount`, and also some syscalls can be dropped if you remove capabilities. Then extending this to other types of configuration, ie more general roles that are defined in terms of multiple types of security. It would then start to make sense if the components that make up this are maybe configurable daemon side, while the roles your container needs are configurable client side. 

(I am talking about this at DevOpsDays London tomorrow actually...)

**Comment by thaJeztah** [comment]
@justincormack yeah, I just did a quick copy paste of some seccomp calls to illustrate, I'm sure I missed some :blush:

My idea is to *merge* the profiles before applying (does that make sense?) i.e. if profile1 marks

profile1;

```JSON
{
    "syscalls": [
        {
            "name": "chmod",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        },
        {
            "name": "chown",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        }
    ]
}
```

profile2;

```JSON
{
    "syscalls": [
        {
            "name": "chmod",
            "action": "SCMP_ACT_ALLOW",
            "args": []
        }
    ]
}
```

Would result in this profile being applied;

```JSON
{
    "syscalls": [
        {
            "name": "chmod",
            "action": "SCMP_ACT_ALLOW",
            "args": []
        },
        {
            "name": "chown",
            "action": "SCMP_ACT_ERRNO",
            "args": []
        }
    ]
}
```

**Comment by justincormack** [comment]
You could make a client side tool that does this very simply, but I don't think this is the right approach. Oh and your example profiles were incompatible, it should not have combined them like that...

**Comment by eparis** [comment]
I generally think it is a bad idea to try to 'merge' profiles. You'll never do it 'right' given only the OCI format information. You can't know exactly what the user wanted. Composing policies is hard, layering policies is 'easy'.

The kernel can layer profiles and it seems to me like exposing that is a good idea. If you load multiple profiles into the kernel they will be executed in order and the 'most denying' action will get executed.  If for a given syscall profile1 has 'errno', profile2 has 'kill', and profile2 has 'trap' the process will get killed.

Layering instead of composing means you can't really do the partial policy described above where you add allows for chmod and chown. It means you can only restrict more than the parent(s). But at least we can rationalize about the end state....

**Comment by justincormack** [comment]
@eparis yes we could allow layering with multiple calls, either layered on top of the default profile or the unconfined profile. The kernel will fail layers that do not have the same default policy, so you can only layer whitelists on whitelists or blacklists on blacklists, which is fine.

**Comment by justincormack** [comment]
It turns out from https://github.com/seccomp/libseccomp/issues/44 that we may want composable profiles but not to merge them.

**Comment by justincormack** [comment]
Another issue is that the format is pretty hard to write (am currently writing some), and there are a lot of limitations of the OCI format, and composition is not one of its strong points. I think the format would have to change to be usable.
